lib/workers/repository/process/vulnerabilities.ts

```diff
@@ -91,51 +97,367 @@ export class Vulnerabilities {
 
-    const vulnerabilities = await this.osvOffline?.getVulnerabilities(
-      ecosystem!,
-      packageDependency.depName!
+      if (!versioningApi.isVersion(depVersion)) {
+        logger.debug(
+          `Skipping vulnerability lookup for package ${packageName} due to unsupported version ${depVersion}`
+        );
+        return [];
+      }
+
+      for (const vulnerability of vulnerabilities) {
+        for (const affected of vulnerability.affected ?? []) {
+          if (
+            this.isPackageVulnerable(
+              ecosystem,
+              packageName,
+              depVersion,
+              affected,
+              versioningApi
+            )
+          ) {
+            logger.debug(
+              `Vulnerability ${vulnerability.id} affects ${packageName} ${depVersion}`
+            );
+            const fixedVersion = this.getFixedVersion(
+              ecosystem,
+              depVersion,
+              affected,
+              versioningApi
+            );
+            if (fixedVersion) {
+              logger.debug(
+                `Setting allowed version ${fixedVersion} to fix vulnerability ${vulnerability.id} in ${packageName} ${depVersion}`
+              );
+              const rule = this.convertToPackageRule(
+                packageFileConfig,
+                dep,
+                packageName,
+                depVersion,
+                fixedVersion,
+                vulnerability
+              );
+              packageRules.push(rule);
+            } else {
+              logger.debug(
+                `No fixed version available for vulnerability ${vulnerability.id} in ${packageName} ${depVersion}`
+              );
+            }
+          }
+        }
+      }
+
+      this.sortByFixedVersion(packageRules, versioningApi);
+    } catch (err) {
+      logger.debug(
+        { err },
+        `Error fetching vulnerability information for ${packageName}`
+      );
+    }
+
+    return packageRules;
+  }
+
+  private sortByFixedVersion(
+    packageRules: PackageRule[],
+    versioningApi: VersioningApi
+  ): void {
+    packageRules.sort((a, b) =>
+      versioningApi.sortVersions(
+        (a.allowedVersions as string).replace(regEx(/[=> ]+/g), ''),
+        (b.allowedVersions as string).replace(regEx(/[=> ]+/g), '')
+      )
     );
-    return this.convertToPackageRule(
-      vulnerabilities ?? [],
-      packageDependency.depName!,
-      ecosystem!
+  }
+
+  // https://ossf.github.io/osv-schema/#affectedrangesevents-fields
+  private sortEvents(
+    events: Osv.Event[],
+    versioningApi: VersioningApi
+  ): Osv.Event[] {
+    const sortedCopy: Osv.Event[] = [];
+    let zeroEvent: Osv.Event | null = null;
+
+    for (const event of events) {
+      if (event.introduced === '0') {
+        zeroEvent = event;
+        continue;
+      }
+      sortedCopy.push(event);
+    }
+
+    sortedCopy.sort((a, b) =>
+      versioningApi.sortVersions(Object.values(a)[0], Object.values(b)[0])
```

Comment:
why such a hacky value access? please replace with normal property access.

Reply:
My challenge here is that the property names vary and the intention is to really to sort by value, no matter how the key is named. E.g., the input can look like this:

```json
[
  { "fixed": "1.7.6" },
  { "introduced": "1.6.0" },
  { "last_affected": "1.7.4" },
  { "introduced": "0.6.0" },
  { "fixed": "0.8.5" },
  { "limit": "1.2.3" }
]
```

I can't really figure out how to do it differently....

Reply:
Interestingly, I cannot find a single vulnerability in the OSV database that uses `limit`.

Reply:
Me neither. I just put it into the snippet above to cover potential keys.

Interesting note about limit from the [examples section ](https://ossf.github.io/osv-schema/#examples-2)in the OSV schema:
> "limit" events are typically not necessary for describing numbered (linear) version ranges and should not be used. They are more useful for git ranges, where it has more implications for the evaluation algorithm.